### PR TITLE
Run Tests in Parallel

### PR DESCRIPTION
**Pull request by harshil21**
Runs the test suite in parallel by using the `pytest-xdist` plugin. In my case, the test suite went from taking ~14 mins to complete to just 3m 28s!
 That's roughly a **75%** improvement!

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -3,6 +3,8 @@ pre-commit
 pytest==7.1.3
 pytest-asyncio==0.19.0
 pytest-timeout==2.1.0 # used to timeout tests
+pytest-xdist==2.5.0  # used to run tests in parallel
+pytest-xdist[psutil]  # used for better auto detection of cpu cores to run in parallel
```
Comment (Bibo-Joshi) [inline requirements-dev.txt:7]:
```
pytest-xdist[psutil]==2.5.0
```
doesn't that work?

File: tests/test_datetime.py [hunk 2]
```diff
@@ -31,8 +31,8 @@
 from tests.conftest import env_var_2_bool
 
 ABSOLUTE_TIME_SPECS = [
-    dtm.datetime.now(tz=dtm.timezone(dtm.timedelta(hours=-7))),
-    dtm.datetime.utcnow(),
+    dtm.datetime.now(tz=dtm.timezone(dtm.timedelta(hours=-7))).replace(second=0, microsecond=0),
+    dtm.datetime.utcnow().replace(second=0, microsecond=0),
```
Comment (Bibo-Joshi) [inline tests/test_datetime.py:35]: why?

File: requirements-dev.txt (hunk 1 above)
Comment (harshil21) [reply requirements-dev.txt:7]: hm, didn't think chaining these would work, but it seems that it does

File: tests/test_datetime.py (hunk 2 above)
Comment (harshil21) [reply tests/test_datetime.py:35]: when initializing pytest, each worker got a slightly different value for this, hence it showed an error saying that it got different items in this list.

Comment (Bibo-Joshi) [reply tests/test_datetime.py:35]: ah, interesting!

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -53,6 +41,7 @@ jobs:
         # 4. Test everything else
         # The first & second one are achieved by mocking the corresponding import
         # See test_helpers.py & test_no_passport.py for details
+        # `--dist loadfile` distributes each test file to a different worker
```
Comment (Bibo-Joshi) [inline .github/workflows/test.yml:44]:
```suggestion
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -65,7 +54,7 @@ jobs:
           no_rate_limiter_exit=$?
           export TEST_RATE_LIMITER='true'
           pip install -r requirements-opts.txt
-          pytest -v --cov --cov-append
+          pytest -v --cov --cov-append -n auto --dist loadfile
```
Comment (Bibo-Joshi) [inline .github/workflows/test.yml:57]:
```suggestion
          # `-n auto --dist loadfile` uses pytest-xdist to run each test file on a different CPU worker
          pytest -v --cov --cov-append -n auto --dist loadfile
```

File: requirements-dev.txt [hunk 5]
```diff
@@ -3,6 +3,7 @@ pre-commit
 pytest==7.1.3
 pytest-asyncio==0.19.0
 pytest-timeout==2.1.0 # used to timeout tests
+pytest-xdist[psutil]==2.5.0  # xdist runs tests in parallel, psutil for better cpu no. detection
```
Comment (Bibo-Joshi) [inline requirements-dev.txt:6]: I re-checked and psutil is not pure-python so that installation may not be available on all setups. Maybe skip the requirement here and instead manually install it in the test workflow? that way we avoid headaches for contribs trying to install the dev reqs

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Sorry for getting back to this late.

The speedup in the CI doesn't seem as big as for you locally, but it's nice nonetheless. Tests seem to run rather stable, too, so let's add it :) Can you also add a note to the contrib guide that tells users how they can use pytest-xdist? Probably also add a warning that this *may* effect the test results in rare edge cases :D

File: .github/workflows/test.yml (hunk 4 above)
Comment (harshil21) [reply .github/workflows/test.yml:57]: can we add comments in that block?

Comment (Bibo-Joshi) [reply .github/workflows/test.yml:57]: Yes, see [here](https://github.com/python-telegram-bot/python-telegram-bot/pull/3267/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88)

**Review by Bibo-Joshi — APPROVED** [review]
LGTM 👍 if there's nothing left from your side, I can merge once the tests pass :)
